P2P improvements (#111)
decentraland/adr: RFC/RFC-14-p2p-adapter.md

A P2P adapter is needed in order to provide a light mechanism for comms (compared to [RFC-5: comms service using websocket](/RFC/RFC-5-ws-comms.md)). This adapter should guarantee the delivery of a message to any peer in the network (or cluster, [real-time clustering of users in Islands (ADR-35)](/adr/ADR-35)).

> The new implementation should solve the problems in the old version[link]:

Reviewer: Should this link point somewhere?

> - When two peers are unable to see each other, it's hard to identify if there is a network failure or a cluster or a bug in the client.

Reviewer: ```suggestion
- When two peers are unable to see each other, it's hard to identify if there is a network failure or they are in different clusters, or a bug in the client.
```

> - Since all peers broadcast information on an ever changing logical mesh (every tick the relay changes), it's almost impossible to measure latency. We also don't know the impact of the amount of packages that are circulating around the network or even how many are been currently relayed. The relay suspension mechanism that prevents the network to be flooded with packages also prevents the stabilization of the mesh.

Reviewer: ```suggestion
- Since all peers broadcast information on an ever-changing logical mesh (every tick the relay changes), it's almost impossible to measure latency. The impact of the number of packages that are circulating around the network or even how many are been currently relayed isn't measurable. The relay suspension mechanism that prevents the network to be flooded with packages also prevents the stabilization of the mesh.
```

> - We don't have a way to measure the impact of the "suspension relay" heuristic, which means we don't know if the available routes are efficent.

Reviewer: ```suggestion
- The impact of the "suspension relay" heuristic can not be measured, which means there is no guarantee that the available routes are efficient.
```

> - There are two unhandled scenarios which may cause messages to not be deliver:

Reviewer: ```suggestion
- There are two unhandled scenarios that may cause messages to not be delivered:
```

>   - if there are two clusters (or more), peers will be completely isolated between each other, no communication using the P2P mesh is possible.
>   - At some point, relay suspension may suspend the only peer that is able to deliver a message. This is a temporal failure, since the next time the mesh is negotiated, the delivery will be resumed. 

Reviewer: ```suggestion
  - if there are two clusters (or more), peers will be completely isolated from each other, and no communication using the P2P mesh is possible.
  - At some point, relay suspension may suspend the only peer that is able to deliver a message. This is a temporal failure, since the next time the mesh is negotiated, the delivery will be resumed. 
```

## Approach

This approach focuses on addressing the problems mentioned above:

>   1. Handle clusters: identifing peers outside of the current mesh, and providing a fallback service to deliver the messages. Optionally (outside of the scope of this RFC), this service could provide hints to the peers in order to connect separated clusters.

Reviewer: ```suggestion
  1. Handle clusters: identifying peers outside of the current mesh, and providing a fallback service to deliver the messages. Optionally (outside of the scope of this RFC), this service could provide hints to the peers in order to connect separated clusters.
```

>   2. Efficient usage of resources: messages are received only once, they are decoded only once, they are relayed using only the most efficent route and thus avoiding flooding the network. There is also no additional expiration ttl, or hops ttl since the route is already determined.

Reviewer: ```suggestion
  2. Efficient usage of resources: messages are received only once, they are decoded only once, and they are relayed using only the most efficient route and thus avoiding flooding the network. There is also no additional expiration TTL or hops TTL since the route is already determined.
```

## The solution

This implementation requires two extra services:

- Routing service: in charge of keeping track of the connections between peers and providing them with routing information for each message (the route the message needs to follow).
- Messaging service: in charge of sending messages between specific peers, when either the connection between two peers is cut, or when the target is in another cluster.

Note: *As defined in [ADR-81](/ADR/ADR-81-minimum-comms-transport.md), each peer will know the ids of the peers around them*

The basic idea of this implementation is for peers to connect randomly to a subset of the others forming a mesh, and reporting their connections to the routing service. The routing service will build the routing information needed for each peer to distribute messages to every other. Each peer will receive from the routing service the minimal set of paths that cover all the connected peers. When a peer needs to deliver a message, it will use the paths provided by the routing service, and if the conection to the neighbour fails, it will fallback to the messaging service for all the pending peers.

> Although this implementation requires a messaging service, it tries to maximize the usage of P2P connections. Also, each peer needs to decode the package to parse it but they don't make changes to it, so there is no cost associated to the enconding of each package.

Reviewer: ```suggestion
Although this implementation requires a messaging service, it tries to maximize the usage of P2P connections. Also, each peer needs to decode the package to parse it but they don't make changes to it, so there is no cost associated with the encoding of each package.
```

### 3. The routing service creates a routing table for each peer

> It is a list of all the needed paths so all the network can be notified by the message. As every message is a broadcast, each peer needs to know the smallest set of paths that covers all the nodes. It will also notify the unreachable nodes, so the peer can do the relay by the Messaging Service. To calculate all the unreachable nodes, each peer will make a diff of the complete list of known peers against all the other peers mentioned in all the paths.

Reviewer: ```suggestion
It is a list of all the needed paths so that the whole network can be notified by the message. As every message is broadcast, each peer needs to know the smallest set of paths that covers all the nodes. It will also notify the unreachable nodes, so the peer can do the relay by the Messaging Service. To calculate all the unreachable nodes, each peer will make a diff of the complete list of known peers against all the other peers mentioned in all the paths.
```

- A given implementation can be optimized by suggesting peers to connect to certain others in order to avoid clustering and minimize messaging service usage. This is out of the scope for this document.
- Since the routing service will know the status of the mesh at all times, it's easy to graph and debug network problems. 

# Risks

> - Potentially the messages may be too big, if they include all the routes. Mitigation: if bandwith becomes a problem, it is possible to encode the message again by removing the current peer from the routing paths, this way the message is small but the process will consume more CPU instead. 

Reviewer: ```suggestion
- Potentially the messages may be too big if they include all the routes. Mitigation: if the bandwidth becomes a problem, it is possible to encode the message again by removing the current peer from the routing paths, this way the message would be smaller but the process will consume more CPU instead. 
```

> - Routing service cost to keep the routes updated. Mitigation: based on metrics we can adjust the parameters: frequency of the routes calculation, the routing algorithm, how we store the graph in memory. 

Reviewer: ```suggestion
- Routing service cost to keep the routes updated. Mitigation: based on metrics we can adjust the parameters: frequency of the routes calculation, the routing algorithm, and how the graph is stored in memory. 
```

- Why not let each peer calculate the best routing table, by broadcasting each peer status to the network?
  - This solution will consume a lot of resources from each client, since the table needs to be updated (or verified) every time there is a change.
  - This solution is somehow more difficult to debug, since routing tables have no source of truth, and different peers may build them differently. 

> - Can we replace the messaging service by providing alternatives routes for broadcasting a message?

Reviewer: ```suggestion
- Can the messaging service be replaced by providing alternative routes for broadcasting a message?
```

  - This won't solve the cluster partitions problem.

>   - By looking at the metrics of the messaging service usage, is possible to conclude the service is not solving a frequent enough problem, and then it can be removed or replaced.

Reviewer: ```suggestion
  - By looking at the metrics of the messaging service usage, it is possible to conclude that the service is not solving a frequent enough problem, and then it can be removed or replaced.
```
